runtime/scripts/results.js

```diff
@@ -0,0 +1,211 @@
+    var results = Numbas.results = /** @lends Numbas.results */ {
+        // items should be accessible through Numbas.results.function, so either write them inside this as key:function pairs, or if necessary as:
+        //var ensure_decimal = math.ensure_decimal = function(n) { ? We need them to be like this to ensure they're accessible from elsewhere, maybe.
+        
+        /** Ensures a string will not cause issues within a csv due to commas, quotes, etc. 
+         * 
+         * @param {string} cell 
+         * @returns {string}
+         */
+        escape_csv_cell : function (cell) {
+            cell = cell + '';
+            if(cell.match(/[,"']/)) {
+                cell = '"'+cell.replace(/\\/g,'\\\\').replace(/"/g,'\\"')+'"';
+            }
+            return cell;
+        },
+        
+        /** Escapes each cell of a list of strings such that each will not cause issues within a csv
+         * 
+         * @param {Array.<string>} cells 
+         * @returns {Array.<string>} 
+         */
+        make_csv_row : function (cells) {
+            return cells.map(Numbas.results.escape_csv_cell).join(',');
+        },
+
+        /** Escapes each cell of a two-dimensional array of strings such that each will not cause issues within a csv
+         * 
+         * @param {Array.<Array.<string>>} rows 
+         * @returns  {Array.<Array.<string>>} 
+         */
+        make_csv : function (rows) {
+            return rows.map(Numbas.results.make_csv_row).join('\n');
+        },
+        
+        /** Creates a csv record of the student's scores and answers, and the maximal scores and correct answers.
+         * 
+         * @returns {string} csv of student's results.
+         */
+        create_results_csv :  function () {
+            //todo: if questions are randomised at all, contain a line mapping to original ordering
+            const randomised = Numbas.results.examIsRandomised();
+            const questionMapping = Numbas.results.questionMapping();
+            let exam = Numbas.exam;
+
+            let header = [''];
+            let expectedAnswers = ['Ideal'];
+            let studentAnswers = ['Student'];
+            let trueOrder = ['True Order'];
+
+            header.push('Total Score');
+            expectedAnswers.push(exam.mark);
+            studentAnswers.push(exam.score);
+            trueOrder.push('');
+
+            for (let questionKey of Object.keys(exam.questionList)) {
+
+                let questionObject = exam.questionList[questionKey];
+                let questionName = questionObject.name||("Question "+questionKey);
+                let trueQuestionNumber = "Question " + questionMapping[questionKey]
+                header.push(questionName);
+                trueOrder.push(trueQuestionNumber);
+                expectedAnswers.push(questionObject.marks);
+                studentAnswers.push(questionObject.score);
+
+                for (let partKey of Object.keys(questionObject.parts)) {

```

Comment:
Having looked at 'allParts', it seems like it would make it more difficult to tell which was a 'part' and which was a 'gap', though that could potentially be bypassed as gap and step have the 'isGap'/'isStep' booleans, so if allParts only covers those 3 we do have 3 distinct cases. I'm currently ambivalent to this change - I think the current format is more human-interpretable, but it does include code duplication.